tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -53,75 +66,110 @@ parseByteString language bytestring = withParser language $ \ parser -> withPars
+newtype B a = B (forall r . (r -> r -> r) -> (a -> r) -> r -> r)
+
+instance Functor B where
+  fmap f (B run) = B (\ fork leaf -> run fork (leaf . f))
+  {-# INLINE fmap #-}
+  a <$ B run = B (\ fork leaf -> run fork (leaf . const a))
+  {-# INLINE (<$) #-}
+
+instance Semigroup (B a) where
+  B l <> B r = B (\ fork leaf nil -> fork (l fork leaf nil) (r fork leaf nil))
+  {-# INLINE (<>) #-}
+
+instance Monoid (B a) where
+  mempty = B (\ _ _ nil -> nil)
+  {-# INLINE mempty #-}
+
+instance Foldable B where
+  foldMap f (B run) = run (<>) f mempty
+  {-# INLINE foldMap #-}
+
+singleton :: a -> B a
+singleton a = B (\ _ leaf _ -> leaf a)
+{-# INLINE singleton #-}
+
+hoist :: (forall x . t x -> t' x) -> Match t -> Match t'
+hoist f (Match run) = Match (fmap f . run)
+{-# INLINE hoist #-}
+
+lookupSymbol :: TSSymbol -> IntMap.IntMap a -> Maybe a
+lookupSymbol sym map = IntMap.lookup (fromIntegral sym) map
+{-# INLINE lookupSymbol #-}
+
+-- | Unmarshal a node
+unmarshalNode :: forall t a .
+                 ( UnmarshalAnn a
+                 , Unmarshal t
+                 )
+  => Node
+  -> MatchM (t a)
+unmarshalNode node = case lookupSymbol (nodeSymbol node) matchers' of
+  Just t -> runMatch t node
+  Nothing -> liftIO . throwIO . UnmarshalError $ showFailure (Proxy @t) node
+{-# INLINE unmarshalNode #-}
 
 -- | Unmarshalling is the process of iterating over tree-sitter’s parse trees using its tree cursor API and producing Haskell ASTs for the relevant nodes.
 --
---   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'unmarshalNode' providing that they have a suitable 'Generic1' instance.
-class Unmarshal t where
-  unmarshalNode
-    :: ( Has (Reader ByteString) sig m
-       , Has (Reader (Ptr Cursor)) sig m
-       , MonadFail m
-       , MonadIO m
-       , UnmarshalAnn a
-       )
-    => Node
-    -> m (t a)
-  default unmarshalNode
-    :: ( Generic1 t
-       , GUnmarshal (Rep1 t)
-       , Has (Reader ByteString) sig m
-       , Has (Reader (Ptr Cursor)) sig m
-       , MonadFail m
-       , MonadIO m
-       , UnmarshalAnn a
-       )
-    => Node
-    -> m (t a)
-  unmarshalNode x = do
-    goto (nodeTSNode x)
-    to1 <$> gunmarshalNode x
-
-instance (Unmarshal f, Unmarshal g, SymbolMatching f, SymbolMatching g) => Unmarshal (f :+: g) where
-  unmarshalNode node = do
-    let lhsSymbolMatch = symbolMatch (Proxy @f) node
-        rhsSymbolMatch = symbolMatch (Proxy @g) node
-    if lhsSymbolMatch then
-      L1 <$> unmarshalNode @f node
-    else if rhsSymbolMatch then
-      R1 <$> unmarshalNode @g node
-    else
-      fail $ showFailure (Proxy @(f :+: g)) node
+--   Datatypes which can be constructed from tree-sitter parse trees may use the default definition of 'matchers' providing that they have a suitable 'Generic1' instance.
+class SymbolMatching t => Unmarshal t where
+  matchers' :: IntMap.IntMap (Match t)
+  matchers' = IntMap.fromList (toList matchers)
```

Comment:
Building an `IntMap` by successive `<>`s is not as efficient as building the whole thing from a list in one step…